File: internal/node/gnunet.go
package node

import (
	"bytes"
	"context"
	_ "embed"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"testing"
	"text/template"
	"time"
)

//go:embed gnunet.conf.tmpl
var gnunetConfTmpl string

type GnunetPeer struct {
	t        *testing.T
	Index    int
	DataDir  string
	ConfPath string
	NS       string
	IP       string
	Port     int
	cmd      *exec.Cmd
}

func NewGnunetPeer(t *testing.T, index int, ns, ip string, port int) *GnunetPeer {
	dataDir := t.TempDir()
	confPath := filepath.Join(dataDir, "peer.conf")

	p := &GnunetPeer{
		t:        t,
		Index:    index,
		DataDir:  dataDir,
		ConfPath: confPath,
		NS:       ns,
		IP:       ip,
		Port:     port,
	}

	// Create necessary directories
	dirs := []string{
		filepath.Join(dataDir, ".cache", "gnunet"),
		filepath.Join(dataDir, "data", "hosts"),
		filepath.Join(dataDir, "run"),
	}
	for _, d := range dirs {
		if err := os.MkdirAll(d, 0755); err != nil {
			t.Fatalf("failed to create dir %s: %v", d, err)
		}
	}

	t.Cleanup(func() {
		if p.cmd != nil {
			exec.Command("ip", "netns", "exec", p.NS, "gnunet-arm", "-c", p.ConfPath, "-e").Run()
			p.cmd.Wait()
		}
	})

	return p
}

func (p *GnunetPeer) WriteConfig() error {
	tmpl, err := template.New("gnunet").Parse(gnunetConfTmpl)
	if err != nil {
		return err
	}

	var buf bytes.Buffer
	err = tmpl.Execute(&buf, p)
	if err != nil {
		return err
	}

	return os.WriteFile(p.ConfPath, buf.Bytes(), 0644)
}

func (p *GnunetPeer) Start() error {
	p.cmd = exec.Command("ip", "netns", "exec", p.NS, "gnunet-arm", "-c", p.ConfPath, "-s", "-L", "DEBUG")
	if err := p.cmd.Start(); err != nil {
		return fmt.Errorf("failed to start gnunet-arm: %v", err)
	}
	return nil
}

func (p *GnunetPeer) ExportHello() (string, error) {
	cmd := exec.Command("ip", "netns", "exec", p.NS, "gnunet-hello", "-c", p.ConfPath, "-e")
	out, err := cmd.Output()
	if err != nil {
		return "", fmt.Errorf("failed to export hello: %v", err)
	}
	return strings.TrimSpace(string(out)), nil
}

func (p *GnunetPeer) ImportHello(hello string) error {
	cmd := exec.Command("ip", "netns", "exec", p.NS, "gnunet-hello", "-c", p.ConfPath, "--import")
	cmd.Stdin = strings.NewReader(hello)
	if out, err := cmd.CombinedOutput(); err != nil {
		return fmt.Errorf("failed to import hello: %v, output: %s", err, string(out))
	}
	return nil
}

func (p *GnunetPeer) WaitCoreConnected(ctx context.Context) error {
	logFile := filepath.Join(p.DataDir, ".cache", "gnunet", "gnunet.log")
	ticker := time.NewTicker(2 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-ticker.C:
			// 1. Try logs
			data, err := os.ReadFile(logFile)
			if err == nil {
				s := string(data)
				if strings.Contains(s, "notification about connection") ||
					strings.Contains(s, "connection established") ||
					strings.Contains(s, "connected to") {
					return nil
				}
			}

			// 2. Try gnunet-statistics
			cmd := exec.Command("ip", "netns", "exec", p.NS, "gnunet-statistics", "-c", p.ConfPath, "-q", "-s", "core")
			out, err := cmd.Output()
			if err == nil && len(out) > 0 {
				s := strings.ToLower(string(out))
				if strings.Contains(s, "neighbour") || strings.Contains(s, "connection") {
					return nil
				}
			}
		}
	}
}

File: testbed/src/node/gnunet.rs
use std::{
    error::Error,
    fs,
    io::Write,
    path::PathBuf,
    process::{Child, Command, Stdio},
    thread,
    time::{Duration, Instant},
};
use tempfile::TempDir;

const GNUNET_TEMPLATE: &str = include_str!("../../templates/gnunet.conf.tmpl");

pub struct GnunetPeer {
    pub index: usize,
    pub data_dir: TempDir,
    pub conf_path: PathBuf,
    pub ns: String,
    pub ip: String,
    pub port: u16,
    child: Option<Child>,
}

impl GnunetPeer {
    pub fn new(index: usize, ns: impl Into<String>, ip: impl Into<String>, port: u16) -> Self {
        let data_dir = tempfile::tempdir().expect("create gnunet tempdir");
        for rel in [".cache/gnunet", "data/hosts", "run"] {
            fs::create_dir_all(data_dir.path().join(rel)).expect("create gnunet dir");
        }
        let conf_path = data_dir.path().join("peer.conf");
        Self {
            index,
            data_dir,
            conf_path,
            ns: ns.into(),
            ip: ip.into(),
            port,
            child: None,
        }
    }

    pub fn write_config(&self) -> Result<(), Box<dyn Error>> {
        let rendered = minijinja::render!(
            GNUNET_TEMPLATE,
            data_dir => self.data_dir.path().display().to_string(),
            ip => self.ip,
            port => self.port,
        );
        fs::write(&self.conf_path, rendered)?;
        Ok(())
    }

    pub fn start(&mut self) -> Result<(), Box<dyn Error>> {
        let child = Command::new("ip")
            .args([
                "netns",
                "exec",
                &self.ns,
                "gnunet-arm",
                "-c",
                self.conf_path.to_str().unwrap(),
                "-s",
                "-L",
                "DEBUG",
            ])
            .spawn()?;
        self.child = Some(child);
        Ok(())
    }

    pub fn export_hello(&self) -> Result<String, Box<dyn Error>> {
        let out = Command::new("ip")
            .args([
                "netns",
                "exec",
                &self.ns,
                "gnunet-hello",
                "-c",
                self.conf_path.to_str().unwrap(),
                "-e",
            ])
            .output()?;
        if !out.status.success() {
            return Err(format!(
                "failed to export hello: {}",
                String::from_utf8_lossy(&out.stderr)
            )
            .into());
        }
        Ok(String::from_utf8_lossy(&out.stdout).trim().to_string())
    }

    pub fn import_hello(&self, hello: &str) -> Result<(), Box<dyn Error>> {
        let mut child = Command::new("ip")
            .args([
                "netns",
                "exec",
                &self.ns,
                "gnunet-hello",
                "-c",
                self.conf_path.to_str().unwrap(),
                "--import",
            ])
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .spawn()?;
        child.stdin.as_mut().unwrap().write_all(hello.as_bytes())?;
        let out = child.wait_with_output()?;
        if !out.status.success() {
            return Err(format!(
                "failed to import hello: {}",
                String::from_utf8_lossy(&out.stderr)
            )
            .into());
        }
        Ok(())
    }

    pub fn wait_core_connected(&self, timeout: Duration) -> Result<(), Box<dyn Error>> {
        let log_file = self.data_dir.path().join(".cache/gnunet/gnunet.log");
        let deadline = Instant::now() + timeout;
        while Instant::now() < deadline {
            if let Ok(data) = fs::read_to_string(&log_file) {
                if data.contains("notification about connection")
                    || data.contains("connection established")
                    || data.contains("connected to")
                {
                    return Ok(());
                }
            }

            let out = Command::new("ip")
                .args([
                    "netns",
                    "exec",
                    &self.ns,
                    "gnunet-statistics",
                    "-c",
                    self.conf_path.to_str().unwrap(),
                    "-q",
                    "-s",
                    "core",
                ])
                .output();
            if let Ok(out) = out {
                let s = String::from_utf8_lossy(&out.stdout).to_lowercase();
                if s.contains("neighbour") || s.contains("connection") {
                    return Ok(());
                }
            }

            thread::sleep(Duration::from_secs(2));
        }
        Err(format!(
            "timeout waiting for GNUnet peer {} core connection",
            self.index
        )
        .into())
    }
}

impl Drop for GnunetPeer {
    fn drop(&mut self) {
        let _ = Command::new("ip")
            .args([
                "netns",
                "exec",
                &self.ns,
                "gnunet-arm",
                "-c",
                self.conf_path.to_str().unwrap(),
                "-e",
            ])
            .status();
        if let Some(mut child) = self.child.take() {
            let _ = child.wait();
        }
    }
}
